### ISSUE TITLE
Add support for prop 224

### ISSUE THREAD

**Issue by Baccount** [micahflee/onionshare#461]
The slugs can be removed because the The first layer of HS descriptor encryption

2.5.1. First layer of encryption [HS-DESC-FIRST-LAYER]
 The first layer of HS descriptor encryption is designed to protect
   descriptor confidentiality against entities who don't know the blinded
   public key of the hidden service.

Here is the design document for Tor's next generation onion service.
https://gitweb.torproject.org/torspec.git/tree/proposals/224-rend-spec-ng.txt

**Comment by Baccount** [comment]
The next generation onion service is expected on December 15th, will OnionShare support before that date? https://trac.torproject.org/projects/tor/wiki/org/teams/NetworkTeam/CoreTorReleases

**Comment by mig5** [comment]
I think OnionShare depends on Stem library for creating the hidden service, so it would depend on whether Stem supports prop224 (which I am struggling to find out).

**Comment by Baccount** [comment]
Opening a ticket on https://trac.torproject.org/projects/tor/report and asking the dev's may be the easiest option.

**Comment by Baccount** [comment]
If I open TorBrowser and have onionshare connect to it, it gets stuck at please wait and never creates the hidden service.

**Comment by Baccount** [comment]
I opened a Ticket asking if stem supports prop 224.
https://trac.torproject.org/projects/tor/ticket/24205#ticket

**Comment by Baccount** [comment]
dgoulet Replied with
“I'm sure it will at some point but for now it can't for ephemeral services until 0.3.3 because prop 224 onion services don't have control port support (#20699).
I would think that Stem can possibly support the creation of persistent service for now. We'll see what atagar decides ;).”

https://trac.torproject.org/projects/tor/ticket/24205#comment:1
@mig5

**Comment by micahflee** [comment]
Yes, I'd love to support this, but it's blocking on stem, which is blocking on prop224 not supporting the control port yet https://trac.torproject.org/projects/tor/ticket/20699.

I think it's less work overall to just wait until the dependencies support the features OnionShare needs before starting work on this. However, as soon as early alpha/dev versions are available to test, I'd like to try to start adding support for OnionShare so it can be ready when the features are stable.

**Comment by Baccount** [comment]
The control port ticket "20699" has been set to merge_ready
@micahflee

**Comment by micahflee** [comment]
We can't make OnionShare support v3 onion services until stem supports it, but I did just made a v3 onion service for the onionshare.org website: http://lldan5gahapx5k7iafb3s4ikijc4ni7gx5iywdflkba5y2ezyg6sjgyd.onion/

When OnionShare checks for updates, it should use the v3 onion service if a new enough version of tor is available, otherwise fallback to the v2 onion service that it currently usees (http://elx57ue5uyfplgva.onion/).

**Comment by mig5** [comment]
Although it's not yet ready in Stem, I've been reading the spec to add v3 Onion Address support to the Control Port.

https://gitweb.torproject.org/torspec.git/tree/proposals/284-hsv3-control-port.txt

These parts will be of importance to how OnionShare requests ADD_ONION:

```
   The "NEW:BEST" option will still return a version 2 address as long as the
   HiddenServiceVersion torrc option default is 2. To ask for a new v3 key,
   this should be used: "NEW:ED25519-V3".
```

```
   Because client authentication is not yet implemented, the "ClientAuth"
   field is ignored as well as "Flags=BasicAuth". A 513 error is returned if
   "ClientAuth" is used with an ED25519-V3 key type.
```

That last one is significant - it means Stealth service setting will need to be disabled/ignored if using v3.

Given that v2 and v3 will co-exist for a while, my suggestion is to make 'Use v3 Onion Services?' a checkbox option in the Settings, rather than automatically try to 'upgrade' to the v3 service in the backend if we detect Stem supports it. Also because by the time v3 comes around, people might be making more use of the 'Persistent URL' feature we have, and so might want to keep their v2 addresses.

**Comment by Baccount** [comment]
mig5 Because V3 will prevent directory crawling, the `slugs` can be removed for the V3 links.

**Comment by mig5** [comment]
I think we can enable by default but if a user has a persistent URL already set, and it's a v2 style private key, then it should continue to use it until the user opts in to upgrade to v3 style via Settings. Otherwise the user loses their v2 private key which might be a big deal for them.

The Stealth thing also remains: if stealth is chosen for use by the user, force v2 (since v3 doesn't support it yet).

Yes I know persistence will be possible (that was why I posted above for future reference, we'll need to request NEW:ED25519-V3 instead of RSA1024)

I am aware of the lack of need for slug (I would argue it could still be useful though in case some future vulnerability is discovered in v3 services that allowed discovery of the address - defense in depth)

**Comment by Baccount** [comment]
@mig5 Maybe making slugs optional for the V3 would be A possible option. You're right it could defend against a possible future exploit but Having a link that's likely 60 characters or more with slugs, may make it more difficult for end users.

**Comment by mig5** [comment]
Surely they will be copying/pasting the URL in any case, especially for v3 :)

**Comment by micahflee** [comment]
I just posted to the tor-dev mailing list asking about when stem will support v3 ephemeral onion services: https://lists.torproject.org/pipermail/tor-dev/2018-January/012841.html

**Comment by Baccount** [comment]
Get Tor 0.3.3 and compile it on macOS instructions 
The source code should be verified against the PGP signature.
`Git clone https://git.torproject.org/tor.git`
`cd tor`
`sh autogen.sh`
`./configure --prefix=/usr/local/tor --with-openssl-dir=/usr/local/ssl --with-libevent-dir=/usr/local/lib && make && make check`
Find tor 0.3.3 at `/Users/"your username"/tor/src/or/tor`
_libevent and OpenSSL should be installed from source._

**Comment by micahflee** [comment]
Have you had any luck trying, @Baccount?

Note that David Goulet responded with:

>This one is on me and I apologize... I haven't took the time to make a proper
control-spec.txt patch from the prop#284.
>
>I'll put this on my priority list so when 0.3.3 goes stable (~ April 15th),
we'll have a up to date control spec for Stem to work with.
>
>Cheers!
>David

So perhaps this issue will be blocked until April. Or, if the "try and see" above works, then perhaps it will just be blocked until Tor 0.3.3 is stable.

**Comment by Baccount** [comment]
I have not tried the commands with Tor 0.3.3, @micahflee are you able to create the V3 onion using information from previous post right above yours?

**Comment by mig5** [comment]
I don’t follow the relevance of 0.3.3 if Stem still doesn’t support it yet, or are we talking about bypassing using Stem now?

Also note that Micah’s test script used NEW:BEST, but as above:

“The "NEW:BEST" option will still return a version 2 address as long as the HiddenServiceVersion torrc option default is 2. To ask for a new v3 key, this should be used: "NEW:ED25519-V3".”

**Comment by Baccount** [comment]
@mig5 have you tested micahflee script with the "NEW:ED25519-V3". option changed?

**Comment by Baccount** [comment]
@micahflee I got it working! Add the onion.py file into a new branch and compile it. Then replace the tor.real with https://github.com/micahflee/onionshare/files/1662660/tor.zip and connect with bundled tor and start a share and it should work. Thank you @mig5 for the helpful infomation
[onion.py.zip](https://github.com/micahflee/onionshare/files/1666763/onion.py.zip)

**Comment by mig5** [comment]
Confirmed on Debian Jessie with Stem 1.6.0 and Tor 0.3.3-alpha that I could create a v3 onion using Micah's script:

```
user@debian-8-dvm:~$ python3 onion.py 
Tor version: 0.3.3.1-alpha
Creating ephemeral onion service
ab4vbqkaaecb53qk22zmlgg6fvv6d6lxr2xck7e3rz6da5uvbbjmmmyd.onion
```

This is great, it implies that we won't have to wait for new Stem at all. We can start doing:

1) version checks of Tor to see if we can attempt a v3 onion
2) checks to see if we can safely make a v3 or keep using the user's persistent v2, maybe with a hint that they can 'upgrade' to persistent v3 onions
3) checks to ignore the Stealth option if requested (since v3 doesn't yet support it)
4) perhaps decide if we want to have an optional 'force v2' setting if for some reason the user wants it (we could re-use this setting silently if a v2 persistent URL is in use)

I'll start on all this very soon if no-one else does.

**Comment by Baccount** [comment]
@mig5 after testing creating V3 onion services in onionshare, it seams that the process is multiple times faster than creating V2's.

**Comment by mig5** [comment]
I'm not going to download anyone's compiled OnionShare for security reasons but I take your word for it :)

I'm working on logic in onion.py to take different actions depending on whether the Tor version support v3 and whether any saved private key is 1024 bits or not (to detect v2 vs v3)

**Comment by Baccount** [comment]
I made a quick video showing the connection times for V2 VS V3. I agree you should never download possibly backdoor app :) 
https://www.icloud.com/iclouddrive/0O1Elsbj_vqEhda8mahG4g4YQ#example

**Comment by maqp** [comment]
https://lists.torproject.org/pipermail/tor-dev/2018-January/012841.html
would appear to be down. Could someone please post @micahflee 's code to this thread/gist/pastebin

**Comment by mig5** [comment]
I'm afraid I don't have my copy of it now, I had loaded it in a Qubes DispVM

I took a stab at this in https://github.com/mig5/onionshare/tree/prop224 and everything is working, except for persistent URLs - for some reason the base64-encoded v3 private key gets rejected with "http status 400 ("Invalid HS descriptor. Rejected.") response from dirserver '185.234.218.44:29736'. Malformed hidden service descriptor?" apparently an issue with decoding, I guess, despite the fact that it is just using the private key returned from the Stem object..

Once again hitting the limit of my skill so maybe others can work it out. Besides if you end up pre-generating the keys rather than rely on what's returned from Stem, as per @maqp suggestion in #572, then maybe that'll be a non-issue.. good luck!

**Comment by Baccount** [comment]
@mig5 
When I compile and run your branch "Prop 224" I get this traceback when the app opens.
```
Traceback (most recent call last):
  File "onionshare-pyinstaller", line 34, in <module>
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/PyInstaller/loader/pyimod03_importers.py", line 631, in exec_module
    exec(bytecode, module.__dict__)
  File "onionshare_gui/__init__.py", line 22, in <module>
    from .alert import Alert
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/PyInstaller/loader/pyimod03_importers.py", line 631, in exec_module
    exec(bytecode, module.__dict__)
  File "onionshare_gui/alert.py", line 22, in <module>
    from onionshare import common
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/PyInstaller/loader/pyimod03_importers.py", line 631, in exec_module
    exec(bytecode, module.__dict__)
  File "onionshare/__init__.py", line 24, in <module>
    from .onion import *
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/PyInstaller/loader/pyimod03_importers.py", line 631, in exec_module
    exec(bytecode, module.__dict__)
  File "onionshare/onion.py", line 24, in <module>
    from Crypto.PublicKey import RSA
ModuleNotFoundError: No module named 'Crypto'
[10830] Failed to execute script onionshare-pyinstaller

```

**Comment by mig5** [comment]
@Baccount it introduced a new dependency pycrypto that you need to install (pip3 install pycrypto)

**Comment by Baccount** [comment]
Thanks, @mig5 it works now.

**Comment by maqp** [comment]
@mig5 

>for some reason the base64-encoded v3 private key gets rejected

https://github.com/torproject/tor/blob/master/src/common/crypto_ed25519.h#L28 Mentions the following:

    /** Note that we store secret keys in an expanded format that doesn't match
    * the format from standard ed25519.  Ed25519 stores a 32-byte value k and
    * expands it into a 64-byte H(k), using the first 32 bytes for a multiplier
    * of the base point, and second 32 bytes as an input to a hash function
    * for deriving r.  But because we implement key blinding, we need to store
    * keys in the 64-byte expanded form. */

It appears to use a format called "[Base64 blob](https://github.com/torproject/tor/blob/b189c613bb27695753761d0851c17f521e8f00de/src/or/control.c#L4850)".

**Comment by mig5** [comment]
That's a good find, but what confuses me is that we're just storing the PrivateKey response sent from control.c itself. You would expect it to send back the key in a format that Tor itself thinks it can work with (but maybe not!)

I dare say this could be worked around if you implement the key generation 'by hand' in onion.py like in the other PR, but it's a bit weird how it doesn't already work..

**Comment by maqp** [comment]
Ha!

Luckily Tor [tests with Python](https://github.com/torproject/tor/blob/8e84968ffbf6d284e8a877ddcde6ded40b3f5681/src/test/ed25519_exts_ref.py).

I converted their code to Python3 so we now have a working key generation + expansion function:

    import hashlib
    import binascii
    import os

    b = 256

    def bit(h, i):
        return (h[i // 8] >> (i % 8)) & 1

    def encodeint(y):
        bits = [(y >> i) & 1 for i in range(b)]
        return b''.join([bytes([(sum([bits[i * 8 + j] << j for j in range(8)]))]) for i in range(b // 8)])

    def H(m):
        return hashlib.sha512(m).digest()

    def expandSK(sk):
        h = H(sk)
        a = 2 ** (b - 2) + sum(2 ** i * bit(h, i) for i in range(3, b - 2))
        k = b''.join([bytes([h[i]]) for i in range(b // 8, b // 4)])
        assert len(k) == 32
        return encodeint(a) + k

    secretKey = os.urandom(32)
    
    expandedSecretKey = expandSK(secretKey)

    key_data = base64.b64encode(expandedSecretKey).decode()

    response = tor.controller.create_ephemeral_hidden_service(ports={80: 5000},
                                                              key_type='ED25519-V3',
                                                              key_content=key_data,
                                                              await_publication=True)

Works like a charm!

**Comment by mig5** [comment]
Thanks so much for this @maqp . I have a working implementation of Onionshare that factors in your features. It generates the private key for both v2 and v3 using your examples, favours v3 if the version of Tor supports it, and I am able to successfully re-share using a persistent URL in both cases. 

I'm going to polish a few little things off and then make a pull request for @micahflee.

**Comment by maqp** [comment]
@mig5 It seems there's a problem with repeated descriptor publishing. I posted the following code that reproduces the bug along with example output to #tor@OFTC but no reply yet: https://gist.github.com/maqp/85452a768b169aa8e2041db2c61ccf67

**Comment by mig5** [comment]
hmm ok, I didn't run into any issue with your code above, interesting..

**Comment by maqp** [comment]
If OnionShare is ready enough, try setting the URL persistent and repeatedly starting and stopping sharing again once the URL is reachable. (The error seems to take 4..7 connections to appear.)

**Comment by mig5** [comment]
Ah ok, I only tested re-sharing twice...

**Comment by mig5** [comment]
@maqp try as I might, I can't reproduce the UPLOAD_REJECTED issue with my branch in #575. Which I guess is a good thing!

**Comment by maqp** [comment]
@mig5 I did some info-level logging with the gist code above. The log had `Malformed hidden service descriptor?` message which lead me to [this bug](https://trac.torproject.org/projects/tor/ticket/23662). So apparently the descriptor needs to be updated every time. I'm not sure if that not happening is a bug in Tor 0.3.3, or if the rev counter is a new feature Stem1.6 doesn't handle.

**Comment by mig5** [comment]
Nice find @maqp - yesterday Micah and I both experienced issues when migrating several v3 onion private keys between servers, and found the same 'Malformed hidden service descriptor' would also occur on the new server, unless the cache was also migrated. After an hour or two it recovered by itself - I strongly suspect this is related to the same rev counter issue and 're-uploading' a pre-generated private key somehow.. v3 feeling a bit brittle still

**Comment by Baccount** [comment]
Opening a ticket on Trac https://trac.torproject.org/ may be the best option if this is a Tor bug. @mig5

**Comment by mig5** [comment]
Probably forget about my prop224 branch @Baccount, I'm not likely to do any more work on it and will stick to just trying to fix smaller bugs

**Comment by maqp** [comment]
@mig5 
I wouldn't mind my chat software to be able to reuse the private key after that hour or two you said (much better than having to create new key pairs every time between all contacts). Did that happen without any additional code, or could you tell how you migrated the cache?

Another thing about the v3 Onion Service creation that bothers me is the [key masking](https://www.youtube.com/watch?v=Di7qAVidy1Y#t=33m24s) that  [part of the testing code](https://github.com/torproject/tor/blob/8e84968ffbf6d284e8a877ddcde6ded40b3f5681/src/test/ed25519_exts_ref.py#L120) does. The idea according to the video above is that the ED25519 is used to sign subkeys that sign the Onion Service descriptor. This solves the issue with .onion site crawling. However, I'm not sure at all Stem 1.6 does all that even with the latest version of Tor, of if Tor's testing code above could do all that.

**Comment by mig5** [comment]
Hi @maqp in this case it was just a 'traditional' Onion Service being migrated, so I'm talking about rsyncing /var/lib/tor to the other server in an attempt to preserve the cache data, which seemed to help. So there seemed some correlation with preserving cache and being able to re-upload the same private key somehow. Obviously not a long term solution as this shouldn't be necessary.

Related to Onionshare, we have a ticket about persisting cache when using bundled Tor, rather than re-initialise fresh cache each time, which certainly would also speed up persistent URLs when someone implements that down the track. When all those slightly-unrelated features converge together, it'll be nice :)

**Comment by maqp** [comment]
@mig5 Hmm, I see. It's probably best to wait until Tor and/or Stem come up with solution before proceeding. [The ticket](https://trac.torproject.org/projects/tor/ticket/23662) has been moved to [0.3.4 milestone](https://trac.torproject.org/projects/tor/milestone/Tor%3A%200.3.4.x-final) indicating it probably won't be fixed anytime soon. I'll try to monitor the situation evenly unevenly at least on

* [Prop224 tickets](https://trac.torproject.org/projects/tor/query?status=accepted&status=assigned&status=closed&status=merge_ready&status=needs_information&status=needs_review&status=needs_revision&status=new&status=reopened&keywords=~prop224&group=status&col=id&col=summary&col=owner&col=type&col=priority&col=component&col=version&order=priority)
* [Atagar's dev blog](http://blog.atagar.com/)
* [Stem unreleased features](https://stem.torproject.org/change_log.html#unreleased)
* [Tor blog](https://blog.torproject.org/)

**Comment by Baccount** [comment]
@maqp explaining  to the Tor devs in the ticket **#23662** that your expecting the issues with implementation of the V3 onion service your app may get the bug prioritized.

**Comment by maqp** [comment]
@Baccount That's a good point. However, as there might be other issues regarding the key blinding and subkeys, I'd rather first see if Stem upgrade (my release has to wait for anyway) fixes the bug by handling descriptor updates. I've just emailed Atagar about Prop284 merging, so I'll know more about the matter soon.

**Comment by Baccount** [comment]
@mig5 @maqp I opened a ticket on Trac and Atagar responded.
```
Hi David. Thanks, and in general I'd agree with you but there's been enough interest in v3 Onion Service support I'd be fine making this into the tracking issue. Adjusting the ticket title to reflect that.
As mentioned on the list I probably won't get to this for a week or two.
```
@maqp Sorry I opened the ticket on Trac before I read your comment on Github.
https://trac.torproject.org/projects/tor/ticket/25124#comment:2

**Comment by maqp** [comment]
@Baccount No worries. However, can you point the Trac ticket's link to https://github.com/micahflee/onionshare/issues/461#issuecomment-361021213 instead of the current one that doesn't address the bug itself.

**Comment by Baccount** [comment]
@maqp just fixed it. https://trac.torproject.org/projects/tor/ticket/25124#comment:3

**Comment by Baccount** [comment]
@maqp Atagar opened a new ticket that may be related to onionshare.
https://trac.torproject.org/projects/tor/ticket/25139

**Comment by maqp** [comment]
@Baccount Interesting to see if that will fix the bug.

Meanwhile, I modified the Gist code above (new [here](https://gist.github.com/maqp/4817a4d1e71a27536ab9810d75e7b6c1)) to perform loop and double the descriptor publish interval every time UPLOAD_REJECTED appears, starting with 60 seconds. Look at the results at the bottom ([link](https://gist.github.com/maqp/4817a4d1e71a27536ab9810d75e7b6c1#file-onion_service_v3-py-L369)) With 3840s delays (a little over an hour), there has yet to happen a single rejection. I'll notify if this changes, but for now it looks like directory services have some rate limiting, or descriptors have lifetime of ~an hour.

This isn't really useful for instant messaging systems like Ricochet or TFC as users shouldn't have to wait for an hour to be able to login e.g. in cases where client crashes just after service is established.

**Comment by Baccount** [comment]
@maqp very cool, lets see what @mig5 says and see if this workaround will work.

**Comment by mig5** [comment]
OnionShare is in the same boat as Richochet/TFC in that we can't wait that long before being able to re-use a private key.

I think there is too much 'mystery' here, we should probably ask the Tor devs explicitly whether this is the issue.

The Tor devs are pre-occupied by 'v3 isn't ready in Stem yet' when we ask - that shouldn't matter for us, because the spec about the control port indicates nothing really 'changes' in terms of how OnionShare uses Stem (the ADD_ONION/DEL_ONION don't change in terms of how we use it).

Instead, this is all about descriptor TTLs and issue when re-uploading before some sort of TTL has expired, and it only affects v3. Nor is it unique to Stem or Onionshare: it occurs even with rsyncing the private key to another server and starting Tor there (after stopping it on the source server). Let's find out if this is on purpose or an oversight.

**Comment by maqp** [comment]
@mig5 

>  the spec about the control port indicates nothing really 'changes'

This is what I'd like to confirm. If `create_ephemeral_hidden_service` would take the expanded signing key, and use that to sign a subkey in v2 spec, Stem 1.6 would already always work. But as per Dingledine's talk, this is a new feature. v3 Onion Services would work consistently if Tor 0.3.3 handled subkey generation when long term signing key is fed through ADD_ONION. However, that can't be the case because we're passing the long term signature verification key, not the long term signing key.

So what I worry is currently happening in my approach above, is Stem 1.6 passes the long term ED25519 verification key inside hidden service descriptor, which allows .onion site crawling. Tor doesn't warn about it because the two keys are indistinguishable 32-byte strings. 

What I'm quite sure will happen is, Stem 1.7 will handle the logic of creating subkeys, signing them with private key we passed it (so this will be an entirely new feature but it's so vastly different I need to dig into specs and confirm it's indeed the case), and send it through ADD_ONION. Stem will also bundle increasing rev counter number and thus removes the issue with UPLOAD_REJECTED.

So while it "works already", it doesn't work as it should because Stem passes wrong data via ADD_ONION. Descriptor TTL isn't the issue if it's solved by descriptors containing working rev counter.

Regarding the server, it's most likely an additional bug, as restarting Tor and Onion Service on server that doesn't use Stem should have handled descriptor update independently. So yes I agree we should pass that as a separate issue that might or might not be related.

**Comment by mig5** [comment]
Nice @maqp, I bow to your knowledge (and I somehow entirely missed that video about the use of subkeys, which I didn't realise was how it was meant to work - probably I got lost in what is becoming a big thread :) )

**Comment by maqp** [comment]
@mig5, @Baccount : Atagar has posted some updates to his [dev blog](http://blog.atagar.com/).

**Comment by maqp** [comment]
@micahflee, @mig5, @Baccount: I just realized persistent URL and rev counter might be a problem in live distros like Tails, where that data disappears between reboots. The current version of Tails (3.5) comes with OnionShare 0.9.2 that does not have the setting for persistent URLs, so I'm not sure this has yet been considered.

When persistence is set for Tails, the OS lets user store data in

* $HOME/Persistent
* GnuPG keyring
* $HOME/.purple (Pidgin+OTR data)
* $HOME/.thunderbird
* GNOME keyring

I think `$HOME/Persistent/onionshare` would be the best place to store the long term private key, as well as any rev counter Stem 1.7 makes available. Then again, Tails developers might be interested in adding another check box for OnionShare so it's worth asking.

**Comment by micahflee** [comment]
@mapq Oh that's a good point, and thanks for noticing. Right now OnionShare stores all of its data in ~/.config/onionshare, including persistent keys. I think the "proper" way for Tails to handle this is to add a new persistence setting for OnionShare, and if that setting is enabled then just store that data too. 

But I think maybe this should be discussed in a separate issue -- perhaps even an issue on the [Tails bug tracker](https://labs.riseup.net/code/projects/tails/issues?query_id=108), since I think this would be more of a Tails change than an OnionShare change.

**Comment by maqp** [comment]
@micahflee Yes, I agree it's probably best to create an issue to Tails bug tracker.

@mig5 I'm not sure what I was thinking when in [this](https://github.com/micahflee/onionshare/issues/461#issuecomment-362974119) comment above I wrote about passing verification key to Stem, when even my example code contradicts that. I must have had my head occupied by the onion URL deriving thing, that uses the signature verification key. I've edited the comment to give more accurate explanation about what is probably going on.

**Comment by Baccount** [comment]
@mig5 @maqp The latest dev stem supports V3 Onionservices, thanks atagar.
https://gitweb.torproject.org/stem.git/commit/?id=53e73ad

**Comment by maqp** [comment]
@Baccount Just ran the stress test script using `Tor 0.3.3.2-alpha-dev` and `stem-1.6.0.dev0`. I am still able to launch the ephemeral service only ~5 times before the UPLOAD_REJECTED error is raised.

On a side note for anyone wondering, the Stem dev build can be installed with
`python3.6 -m pip install git+https://git.torproject.org/stem.git`

**Comment by Baccount** [comment]
@maqp it looks like the Tor devs will discus this problem in there Rome meeting.
`persistent v3 onion service revision counters, self-generated onion service addresses ([#25124](https://trac.torproject.org/projects/tor/ticket/25124))`
https://trac.torproject.org/projects/tor/wiki/org/meetings/2018Rome/AgendaIdeas

**Comment by maqp** [comment]
@mig5 As per Teor's [comment](https://trac.torproject.org/projects/tor/ticket/25124#comment:12) in the upstream ticket, there's no onion address crawling risk when passing the expanded private key to `create_ephemeral_hidden_service` -- key blinding is handled by Tor automatically regardless of Stem. so that's finally cleared.

**Comment by Baccount** [comment]
@maqp Do you think onshore should implement V3 onions without the Persistent option and add that feature in the future when the Tor devs come up with a solution? mig5's branch should accomplish most if not all of this. "Non Persistent to V3 onion addresses"
https://github.com/mig5/onionshare/tree/prop224

**Comment by maqp** [comment]
@Baccount I don't see why it couldn't. The UI for now just needs to disable the possibility to set persistent address and stealth onion service when using v3 Services. IMO v3 should be the default so the Settings could say e.g.

☑ Use v3 onion services
(unchecking will use old, less secure v2 onion services)

☐ Use a persistent address (v2 only)

☐ Create stealth onion services (v2 only)*

When v3 is checked, the two settings should be greyed out.

*[prop224 currently does not specify stealth client authorization.](https://trac.torproject.org/projects/tor/ticket/20742)

**Comment by Baccount** [comment]
@maqp Nick m just merged a patch that may resolve the issue with the rev counter.
https://gitweb.torproject.org/tor.git/commit/?id=5804ccc9070dc5443f1c6ce565dbf17572812764
And arma said
`Onion service rev counters are useless and actually harmful for scalability`
https://trac.torproject.org/projects/tor/ticket/25552

**Comment by maqp** [comment]
@Baccount I'm still getting the UPLAOAD_REJECTED. This is most likely because HSDirs are still using the latest stable or at best, the lastest alpha release. So it may take days, weeks or months before we know it was fixed.

**Comment by Baccount** [comment]
Ok, thank you for checking @maqp .

**Comment by Baccount** [comment]
@maqp  it looks like tero said the rev counter will be removed.
```
In #25552, we decided to remove revision counter checks on HSDirs.
The remaining issue in this ticket is about generating onion addresses using your own entropy.
That's a feature that Stem could support, or you could write a simple script to do it, or we could add a command-line option to tor.
It's unlikely to happen in Tor, so please write your own script, or submit a patch to Stem.
```
[#25552](https://trac.torproject.org/projects/tor/ticket/25552)

**Comment by maqp** [comment]
@Baccount Okay, then it's just the matter of waiting until the next stable version is deployed across HSDIrs. In the meantime, it wouldn't be a bad idea to create a pull-request to stem for the key blob generator for both v2 and v3 services. This would help us double-check the implementation is correct. It would also be cross-platform and helpful for other developers. What makes this simple is both `cryptography` and `pynacl` are listed in [stem dependencies](https://gitweb.torproject.org/stem.git/tree/requirements.txt).

**Comment by maqp** [comment]
@Baccount, @micahflee, @mig5 
Related to this and https://github.com/micahflee/onionshare/issues/677, it looks like [this](https://gitweb.torproject.org/tor.git/commit/?id=a4fcdc5decfe60bbd95aee2e5586e90c40b73225) Tor commit completely broke v3 onion services with stem. You can find the bash script that reproduces the bug [here](https://gist.github.com/maqp/5ea6cf41a836c290b22ec302213d8a00). By default that script uses the latest working commit*. Make sure to comment line 22 and uncomment line 23 before running to see the first dysfunctional commit being unable to start the onion service. I'll try to create a ticket to Tor's trac soon, meanwhile could you check if you can reproduce the problem?

*Working in the sense it only had the problem of raising the UPLOAD_REJECTED after 2..10 descriptor uploads.

**Comment by mig5** [comment]
I'm going to close this as v3 support has been merged in (yay).

We have sub-ticket(s) for supporting persistence or Client Auth for v3 when it becomes possible, I believe.

**Comment by maqp** [comment]
@mig5 @Baccount Stem 1.7 with documented v3 Onion Service support has been [released](https://stem.torproject.org/change_log.html#version-1-7) :smile:
